### PR TITLE
Prune/remove indentation from JSON property values

### PR DESCRIPTION
**Pull request by ronaldbarendse**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes: https://github.com/umbraco/Umbraco-CMS/issues/2996 (again, but for v8, as the changes from https://github.com/umbraco/Umbraco-CMS/pull/6416 probably weren't correctly merged up).

### Description
While working on PR https://github.com/umbraco/Umbraco-CMS/pull/11805, I noticed we're not only storing verbose property values for the Image Cropper and Media Picker (v3), but a lot of other complex editors also stored redundant data (e.g. empty strings/arrays, whitespace/indented JSON, etc.).

Most issues are caused by converting to/from JSON types and strings over and over again, eventually returning serialized JSON that contains indentation or even converting a `null` value into an empty JSON array again (and storing `[]`).

One very important thing to note regarding property data: whenever the value is `null`, it doesn't insert a row into the `umbracoPropertyData` table. This means not storing empty strings or JSON arrays will save much more database storage then just the few characters of the actual value!

Testing can be done by saving content with different editors from the back-office and inspecting the database values. You can create a document type containing all editors using the code from this Gist: https://gist.github.com/ronaldbarendse/6eda1a0bfee87bdd714409886ecf4219.

The following SQL query returns all property data with the latest versions first:
```sql
SELECT
    cv.nodeId,
    cv.text,
    pt.Alias,
    pd.languageId,
    pd.segment,
    pd.intValue,
    pd.decimalValue,
    pd.dateValue,
    pd.varcharValue,
    pd.textValue
FROM
    umbracoPropertyData pd
    INNER JOIN umbracoContentVersion cv ON cv.id = pd.versionId AND cv.[current] = 1
    INNER JOIN cmsPropertyType pt ON pt.id = pd.propertyTypeId
ORDER BY
    cv.versionDate DESC
```

### CODE REVIEW

**Comment by ronaldbarendse** [comment]
I've tested these changes together with PR #11805 (temporarily merged into the [`v8/spike/prune-testing`](https://github.com/umbraco/Umbraco-CMS/tree/v8/spike/prune-testing) branch, but we can update this PR with the latest `v8/dev` when it's merged) and created 2 components that cleans up and prunes property data of an existing site: https://gist.github.com/ronaldbarendse/b58bcfcdd68440ccdd316903cd5c8a03.

To test the code, create a local release (using the `build/build.ps1` script), update an existing site (using `Update-Package UmbracoCms -Source "path/to/build.out"`, see [docs](https://docs.microsoft.com/en-us/nuget/consume-packages/reinstalling-and-updating-packages#using-update-package)), copy `PrunePropertyDataComponent.cs` (from the Gist) to the site project and:
1. Start and upgrade the site, so you can check whether everything still works;
2. Enable the `PublishCleanupComposer` (by removing the comment in the Gist file) and start the site: this re-publishes all content, empties the content and media recycle bin, deletes all previous content versions and rebuilds the NuCache;
3. If you have a large site and get an exception while Umbraco is writing to the NuCache, configure a larger block size: https://github.com/umbraco/Umbraco-CMS/issues/4998#issuecomment-478101534;
4. Write down the statistics you want to track, e.g. the amount of table rows/data used:
  ```sql
  EXEC sp_spaceused [umbracoPropertyData]
  EXEC sp_spaceused [cmsContentNu]
  ```
5. Now also enable the `PrunePropertyDataComposer` and start the site: this will first re-save all property data using the new changes/optimizations and run the cleanup again;
6. Compare the statistics and notice a reduction is stored property data 🎉 

In the sample site I used (that contained around 2000 content/media items), this reduced the total amount of rows in `umbracoPropertyData` from 67.343 down to 44.150 (65% of the original amount) and the total size of the `cmsContentNu` table (using JSON storage) from 64MB down to 55MB (86% of the original total).

**Review by nikolajlauridsen — COMMENTED** [review]
Only saw one thing that puzzled me a bit, otherwise it looks all good to me and tests out good as well 👍

File: src/Umbraco.Web/PropertyEditors/ImageCropperPropertyEditor.cs [hunk 1]
```diff
@@ -241,17 +239,12 @@ private void AutoFillProperties(IContentBase model)
...
-                                src = svalue,
-                                crops = config == null ? Array.Empty<ImageCropperConfiguration.Crop>() : config.Crops
-                            };
 
-                            property.SetValue(JsonConvert.SerializeObject(json), pvalue.Culture, pvalue.Segment);
+                            property.SetValue(JsonConvert.SerializeObject(new
+                            {
+                                src = svalue
+                            }, Formatting.None), pvalue.Culture, pvalue.Segment);
                         }
```
[the remark this answers is not in the document]
Comment (ronaldbarendse) [reply src/Umbraco.Web/PropertyEditors/ImageCropperPropertyEditor.cs:248]: If the value is just the URL, it's still better to re-save this as a JSON value, so the PVC can just deserialize it and not log errors: https://github.com/umbraco/Umbraco-CMS/blob/5bfab13dc5a268714aad2426a2b68ab5561a6407/src/Umbraco.Infrastructure/PropertyEditors/ValueConverters/ImageCropperValueConverter.cs#L50-L60

**Comment by nul800sebastiaan** [comment]
This accidentally breaks something in DTGE and Skybrud.Umbraco.Elements. That problem is fixed in 8.18.1 with PR https://github.com/umbraco/Umbraco-CMS/pull/12138

**Pull request merged**